### PR TITLE
Add shared base app for game launchers

### PR DESCRIPTION
**Pull request by gasinvein**
Main idea behind this is to join efforts of multiple game launchers (e.g. Steam, Lutris, GameHub, minigalaxy, itch.io , etc) on bundling common game dependencies in one place. This way a launcher can base on this app instead of bundling libraries themselves, and all launchers benefit from from a library added.

This baseapp slightly extends flatpak-builder manifest syntax to support multilib builds by using procedural-generated manifest files, in order to reduce maintenance burden. I.e. a module have to be added only once in `modules/`, its two multilib variants are generated in `generated/`.

CC @nanonyme @RobLoach @tim77

### CODE REVIEW

File: org.flathub.GameLauncherBaseApp.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+id: org.flathub.GameLauncherBaseApp
```
Comment (gasinvein) [inline org.flathub.GameLauncherBaseApp.yml]: The app ID is a subject to discuss. Maybe we should pick something else.

**Comment by gasinvein** [comment]
bot, build org.flathub.GameLauncherBaseApp

**Comment by LeandroStanger** [comment]
bot, build org.flathub.GameLauncherBaseApp

**Comment by LeandroStanger** [comment]
bot, build org.flathub.GameLauncherBaseApp

**Comment by RobLoach** [comment]
Seems like a good idea. My only concern would be if dependencies diverge. Don't think that would really be a big issue though, considering we could just keep this targeting the latest version of each.

**Comment by gasinvein** [comment]
@RobLoach Hm, if two (or more) different games have conflicting dependencies, we can't do anything about it regardless it's baseapp or not, right? Multiple versions of the same library can be installed simultaneously as long as they have different soversions, though.

File: modules/wine-deps.yml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+  - name: SPIRV-Headers
+    x-multilib: false
+    buildsystem: cmake-ninja
+    sources:
+      - type: git
+        url: "https://github.com/KhronosGroup/SPIRV-Headers.git"
+        tag: 1.5.1.corrected
+
+  - name: vkd3d
```
Comment (gasinvein) [inline modules/wine-deps.yml]: Should we bundle vkd3d? Seems like it's not universal, vanilla Wine, Lutris wine and Proton use different versions, so probably vkd3d should be bundled alongside wine instead.

**Comment by LeandroStanger** [comment]
bot, build org.flathub.GameLauncherBaseApp

**Comment by TingPing** [comment]
I know you put a lot of work into this but the entire concept seems very anti-flatpak to me.

There is so much contained in here its impossible to reasonably say which app wants what and juggle maintaining it without ever conflicting with the needs of other apps.

**Comment by gasinvein** [comment]
@TingPing Well, game launchers in general, i.e. apps that launch arbitrary other apps, don't fit well with flatpak concept. Still, there are some on Flathub, and there is a demand for adding other. I just want to reduce amount of work needed for each of them.

**Comment by LeandroStanger** [comment]
bot, build org.flathub.GameLauncherBaseApp

**Comment by gasinvein** [comment]
@barthalion Build now fails due to missing 128x128 icon. This baseapp skips appstream check, shouldn't icon check be skipped as well?

**Comment by eszlari** [comment]
@TingPing
> There is so much contained in here its impossible to reasonably say which app wants what and juggle maintaining it without ever conflicting with the needs of other apps.

The Steam flatpak has reached that level of complexity already, it's seems wise to share the burden.

Long term, we want upstream involvement anyway. And the only way to get there, is popularity. Having a mostly working flatpak for Steam and Lutris (I would focus on these two most popular launchers) would help us get there.

It would give flatpak also a strong advantage over snap. They still have not resolved that (https://discourse.ubuntu.com/t/steam-as-a-snap/10733/20), and would nice to grab the mind share before they do.

**Comment by nanonyme** [comment]
@eszlari actually Snap already has one *big* advantage over Flatpak for Steam: Snap supports unsandboxed applications so it can support the new runtime containers (pressure-vessel) as well. Flatpak Steam cannot.

**Comment by eszlari** [comment]
> new runtime containers (pressure-vessel)

According to this presentation by its developer (@smcv), still pretty experimental technology: https://youtu.be/KrbWbBYAolo

**Comment by smcv** [comment]
Yes, pressure-vessel is still quite experimental. We're working on it!

One of the things I definitely *want* to do with pressure-vessel is to make it able to coexist with Flatpak, so that it can launch a new sandbox (in parallel with the one that contains the Steam client) for each game, with that game's preferred runtime environment if that's different.

Initial prototypes for that will probably rely on the Flatpak Steam app having its permissions changed so that it is basically unsandboxed, with arbitrary code execution on the host system, as a big-hammer approach to letting it run bubblewrap directly. However, obviously that's a big step backwards when compared with the privileges the Flatpak Steam app currently has, and it would be better if we can enhance Flatpak so that `flatpak-spawn --sandbox` can do what we need.

This involves changing some assumptions inside Flatpak, and also breaking/changing/fixing some assumptions inside Steam (the IPC between games and Steam currently assumes pids are a flat global namespace, and gets very confused when a game in a separate pid namespace reports "hello, I am process 2"), so it's a significant bit of work.

> Well, game launchers in general, i.e. apps that launch arbitrary other apps, don't fit well with flatpak concept

Yes. At the moment, the Steam client wears a lot of hats: it's a UI (the Store, Library, etc.), it's a download manager, it's a launcher, it's a parent process for games (it doesn't really sandbox them, particularly when not using pressure-vessel, but it does sandbox-adjacent things like acting as a subreaper), it provides its own conceptually Flatpak-runtime-like library stack, ...

I'm sure other game launchers have many of the same roles, particularly those that were developed for Windows and ported to Linux afterwards.

On one hand, the thing that would make most sense in a Flatpak world would be for each game to behave as though it was a separate Flatpak app. This is why pressure-vessel uses directories like `~/.var/app/com.steampowered.App440` in its current prototype of unsharing the home directory.

On the other hand, the developers of these game launchers already have distribution channels, CDNs, packaging formats, app-stores etc. that they are happy with, and are unlikely to be willing to change those for Flatpak's benefit - particularly if (like Steam) they are cross-platform, with a majority of their users still using Windows despite our best efforts.

https://github.com/ValveSoftware/steam-runtime/blob/master/doc/goals.md and https://github.com/ValveSoftware/steam-runtime/blob/master/doc/possible-designs.md summarize some of my thinking about the future of pressure-vessel. Comments welcome (perhaps in the form of MRs for that repository).

https://github.com/ValveSoftware/steam-runtime/blob/master/doc/possible-designs.md#flatpak--2018-ld_library_path-scout-runtime is how the current Flatpak Steam app works.

https://github.com/ValveSoftware/steam-runtime/blob/master/doc/possible-designs.md#flatpak--pressure-vessel-in-parallel is the idea I describe above.

**Comment by eszlari** [comment]
@smcv Why not sub-sandboxes? Flatpak needs them anyway for browsers (https://discourse.gnome.org/t/sandboxing-portal/1651/10).

**Comment by smcv** [comment]
I do intend to use sub-sandboxes, but those will need some enhancements to be able to do what pressure-vessel does.

**Comment by plata** [comment]
I'm looking for something like this for Phoenicis PlayOnLinux 5. For my understanding: Wouldn't it be possible to implement this with runtime extensions such that we could pick the ones we need while not having to worry about every software update?

**Comment by gasinvein** [comment]
Closing in favor of #1830, which has limited scope.

**Pull request closed without merging**